---
title: "12/13 update"
author: "Linda Gai"
date: "12/13/2018"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE, message=FALSE,warning=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{r, message=FALSE,warning=FALSE}
library(dplyr)
library(ggplot2)
```

## R Markdown

## Cleaning vcf

The raw vcf `/dcl01/beaty/data/gmkf/euro/vcfs/filtered/8q24.recode.vcf` was cleaned by:

1. Removing all tri-alellic SNPs

2. Individuals with Mendelian errors

3. Phased using BEAGLE 4.0 (i.e., half calls were not removed).

4. Monomorphic SNPs were removed.

TODO: filter to MAF cut-off to 0.05 or 0.01 (is MAF possibly captured in functional annotation information)?

## Filtering

The ultimate goal was to run all 3 tests on the same datasets to compare results. Since rvTDT uses allele frequencies in the population to weight SNPs in the case-parent trios, I first filtered to all SNPs that had allele frequences in Europeans available from 1000Genomes from Aug 2015 (in the Annovar Report). I then created datasets with 3 different types of filters.

### 1. Filtered by functional annotation information (848 SNPs)
I filtered to SNPs with scores (in the ANNOVAR report) of CADD > 10, GWAVA > 0.4, EIGEN >4, using recommended cut-offs from literature, or, in the case of EIGEN, a cut-off used in at least one paper. I used these scores because the other annotation scores (SIFT, PolyPhen) were completely missing from the ANNOVAR report. Unfortunately, there was high degree of missingness in each of the scores, so this approach may have excluded some posibly causal SNPs if no annotation score was available for them.

```{r, echo=FALSE,message=FALSE,warning=FALSE}
filepath_genotypic_tdt_results<-"/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/8q24_genotypic_tdt_results.txt"
genotypic_tdt_results<-read.table(filepath_genotypic_tdt_results,header=TRUE)

filepath.annovar<-"/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/8q24_annovar_report_functional_annnotation.txt"
annovar<-read.table(filepath.annovar,sep="\t",header=TRUE, quote ="")
pos.annotation<-annovar$StartPosition

genotypic_tdt_results <- genotypic_tdt_results %>%
  mutate(annot.pos = (position %in% pos.annotation)) %>%
  mutate(peak.pos = (position > 129875000 & position < 130000000)) %>%
  mutate(both.pos = (position > 129875000 & position < 130000000) & (position %in% pos.annotation) )
```

```{r,message=FALSE,warning=FALSE}
ggplot(genotypic_tdt_results)+
  geom_point(aes(x=position,y=neglogp,color=annot.pos)) +
  labs(title="Genotypic TDT results, high annotation score positions labeled green",
       x="SNP position", y = "-logp")
```

### 2. Filtered to region with peak TDT signal (512 SNPs)

The positions that have been kept have been labeled red in the gTDT plot below.
Score ranges for .

```{r}
ggplot(genotypic_tdt_results)+
  geom_point(aes(x=position,y=neglogp,color=peak.pos)) +
  labs(title="Genotypic TDT results, peak positions labeled green",
       x="SNP position", y = "-logp")
```

### 3. Both 1 and 2 (102 SNPs)

The positions that have been kept have been labeled red in the gTDT plot below.

The results for the 3 tests (rvTDT, RV-TDT, and ScanTrio are given below):

```{r}
ggplot(genotypic_tdt_results)+
  geom_point(aes(x=position,y=neglogp,color=both.pos)) +
  labs(title="Genotypic TDT results, high annotation score positions in peak labeled green",
       x="SNP position", y = "-logp")
```

Other types of filtering to try:

1. Filtering to high gTDT signal positions? There does not seem to much concordance between having a high annotation score and high gTDT signal, and some higher signal positions lie outside the peak.

2. Filtering to only rare variants of MAF <0.05 in 1000 genomes? in the dataset?
Is there correlation between high functional annotation scores and variant rarity?

3. Filtering to smaller windows -- I've only tried filtering to relatively large windows (76-100 SNPs) using RV-TDT.

## Results
```{r,echo=FALSE}

filepath.rvTDT.annotation<-"/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/rvTDT_results_annotation.rds"
filepath.rvTDT.peak<-"/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/rvTDT_results_peak.rds"
filepath.rvTDT.both<-"/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/rvTDT_results_both.rds"

rvTDT.results<-c(filepath.rvTDT.annotation,filepath.rvTDT.peak,filepath.rvTDT.both)

```

#### rvTDT

So far, for these three filtering approaches, I only have results for rvTDT. Interestingly, it seems like the p-values for the SNPs in the peak are slightly *higher* than the p-values for the SNPs in the peak that were ALSO filtered by annotation information.

QUESTION: Might this indicate that some associated/potentially causal SNPs are being filtered out because of the annotation scores? (Many SNPs have missing annotation scores, or else they might just not have high ones.)

The p-values from the gTDTs using kernels also tend to be lower than the gTDTs using linear combinations, with the kernels weighted by the population controls giving the lowest p-values of all.

KEY:

* p_lc_1: p-value from gTDT w/unweighted linear combination of SNPs

* p_lc_maf: p-value from gTDTw/linear combination of SNPs, weighted by MAF

* p_lc_pc: p-value from gTDT w/linear combination of SNPs, weighted by MAF

* p_k_1: p-value from gTDT w/unweighted kernel SNPs, weighted by population controls (i.e MAF in the population)

* p_k_maf: p-value from gTDT w/kernel of SNPs, weighted by MAF

* p_k_pc: p-value from gTDT w/kernel of SNPs, weighted by by population controls (i.e MAF in the population)

#### 1. Filtered by functional annotation information (CADD > 10, GWAVA > 0.4, EIGEN >4, using literature) using cut-offs


```{r}
unlist(readRDS(rvTDT.results[1]))
```

#### 2. Filtered to peak TDT signal


```{r}
unlist(readRDS(rvTDT.results[2]))
```

#### 3. Both 1 and 2


```{r}
unlist(readRDS(rvTDT.results[3]))
```

## Alternative Filtering approach

#### Windows of size 100, filtered to CADD>10, overlap of 0 and 10, in peak and in whole region

#### RV-TDT

Over the summer, I also tried using windows and filtering on RV-TDT. I first filtered to all positions with a CADD score >10, since that annotation score had the most non-missing entries in the region. I then used relatively large windows of 76 or 100 SNPs, with either no overlap or a 10 SNP overlap between the windows.

#### Results

##### Overall findings; highest signal is possibly outside of the peak
Overall, it seems like the SNPs that were the 600-800th SNPs (by position) in the dataset, out of ~1000 had the lowest-p-values. My guess is that these SNPs are possibly around but not necessarily in the peak -- when I ran RV-TDT on the peak, the p-values were considerably greater in the 600-800th SNPs group than in the peak. However, I would need to plot the 600th-800th positions on the gTDT to confirm this.

Adding overlaps changed the p-values somewhat. The 2 smallest p-values in the overlap data set were 0.11 and 0.2 (SNPs ~540-630, 630-720), compared to 0.25 and 0.13 in the non-overlap data set (SNPs 600-700,700-800).

There was also an additional region in both the overlap and non-overlap datasets that had a slightly lower p-value (0.6-0.7). In the the non-overlap dataset, the last 100 SNPs have a slightly lower p-values (~0.6 using CMC-Analytical and BRV-Haplo, ~0.7-0.8 for the rest), and the second-to-last 100 SNPs also have slightly lower p-values (~0.6-7 using VT-BRV-Haplo, T-CMC-Haplo, WSS-Haplo, ~1 for the others).


 
##### P-values differed widely based on weighting scheme in region
The p-values differed quite a lot between the different weighting schemes. 

##### Lowest p-value method used indicator of whether parent transmitted a RV, association between RV carrier status and phenotype
CMC-Analytical and CMC-Haplo gave relatively low p-values for this region (0.16-0.4). The CMC statistics are the simplest, and just use an indicator variable to indicate whether the parent has rare variants and tests for the association between phenotype and rare variant carrier status. So it seems probable/possible that there is some variant in the 600-800th SNP region that might be associated with cleft phenotype, although the test did not reach significance.

##### Highest p-value methods were weighted by the number of (or variance in the number of) variants transmitted per parent
Conversely, BRV-Haplo, VT-BRV-Haplo, and WSS-Haplo gave relatively higher p-values (~0.5-0.75). BRV counts the number of rare variants transmitted, WSS weights each variant site by the number of variants in the parental haplotypes that are not transmitted in the parental off-spring, and VT maximizes the test statistic over the allele frequencies. 

##### QUESTION/TODO: Does this imply that, of the rare variants transmitted, only a small number are associated with the phenotype? Or that parents that transmitted an associated rare variant transmitted only a few rare variants?

##### One method that gave a low and high p-value maximizes test statistic over allele frequencies
The VT-CMC-Haplo gave a fairly high p-value for the 600th-700th SNPs (~0.75) but a relatively low p-value for the 700-800th SNPs (~0.3). 

##### QUESTION/TODO: How to interpret the information in light of the fact that VT maximizes the test statistic over the allele frequencies? VT is implemented using CMC, so I think the low p-value probably comes from the CMC part, rather than the VT part...the CMC-Analytical p-value (0.13) is quite a lot lower than the VT-CMC (~0.3)

##### P-values within peak relatively high
The highest signal in the peak was in the first half of the peak (76 SNPs, p \approx 0.5). But again, the signal was higher in the 600-800th SNPs.

TODO: 

1. Plot the position of the 600-800th SNPs group on the gTDT dataset.

2. Make the below tables easier to read :)

Results are here:

#Whole 8q24 region, no overlap, 100 SNPs per window

```{r,echo=FALSE}
subdirectories<-c("cadd_10_window-size=100_snps_1_pval",
"cadd_10_window-size=100_snps_2_pval",
"cadd_10_window-size=100_snps_3_pval",
"cadd_10_window-size=100_snps_4_pval",
"cadd_10_window-size=100_snps_5_pval",
"cadd_10_window-size=100_snps_6_pval",
"cadd_10_window-size=100_snps_7_pval",
"cadd_10_window-size=100_snps_8_pval",
"cadd_10_window-size=100_snps_9_pval")

directory<-" '/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/cadd_10/window100_no_overlap/'"

for (i in 1:9){
  sub.directory<-subdirectories[i]
  command<-paste0("cat ",directory,sub.directory,"/8q24.pval")
  cat(system(command,intern = TRUE))
  cat("\n")
  cat("\n")
}

```


#Whole 8q24 region, 10 overlap, 100 SNPs per window
```{r,echo=FALSE}
subdirectories<-c("cadd_10_window-size=100_snps_1_pval",
"cadd_10_window-size=100_snps_2_pval",
"cadd_10_window-size=100_snps_3_pval",
"cadd_10_window-size=100_snps_4_pval",
"cadd_10_window-size=100_snps_5_pval",
"cadd_10_window-size=100_snps_6_pval",
"cadd_10_window-size=100_snps_7_pval",
"cadd_10_window-size=100_snps_8_pval",
"cadd_10_window-size=100_snps_9_pval")

directory<-" '/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/cadd_10/window100_10_overlap/'"

for (i in 1:9){
  sub.directory<-subdirectories[i]
  command<-paste0("cat ",directory,sub.directory,"/8q24.pval")
  cat(system(command,intern = TRUE))
  cat("\n")
  cat("\n")
}

```

#TDT peak, no overlap, 100 SNPs
```{r,echo=FALSE}
subdirectories<-c("cadd_10_peak_window-size=100_snps_1_pval",
                  "cadd_10_peak_window-size=100_snps_2_pval")

directory<-"'/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/cadd_10/peak_no_overlap/'"

for (i in 1:2){
  sub.directory<-subdirectories[i]
  command<-paste0("cat ",directory,sub.directory,"/8q24.pval")
  command
  cat(system(command,intern = TRUE))
  cat("\n")
  cat("\n")
}
```

#TDT peak, no overlap, 76 SNPs
```{r,echo=FALSE}
subdirectories<-c("cadd_10_peak_window-size=76_snps_1_pval",
                  "cadd_10_peak_window-size=76_snps_2_pval")

directory<-" '/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/cadd_10/peak_no_overlap/'"

for (i in 1:2){
  sub.directory<-subdirectories[i]
  command<-paste0("cat ",directory,sub.directory,"/8q24.pval")
  command
  cat(system(command,intern = TRUE))
  cat("\n")
  cat("\n")
}
```

The graphs of the 2 regions in 8q24 with the lowest p-values are given below:

```{r,echo=FALSE}
filepath.10.overlap.map.6<-"/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/cadd_10/window100_10_overlap/cadd_10_window-size=100_snps_6.map"
overlap.map.6<-read.table(filepath.10.overlap.map.6,header=FALSE,stringsAsFactors = FALSE)
head(overlap.map.6)
filepath.10.overlap.map.7<-"/Users/lindagai 1/Documents/classes/4th year/Research/8q24_project/Update 12:13/cadd_10/window100_10_overlap/cadd_10_window-size=100_snps_7.map"
overlap.map.7<-read.table(filepath.10.overlap.map.7,header=FALSE,stringsAsFactors = FALSE)
overlap.map.7$V2
overlap.pos<-unique(c(overlap.map.6$V2,overlap.map.7$V2))
overlap.pos

genotypic_tdt_results <- genotypic_tdt_results %>%
  mutate(overlap.pos = (snp_name %in% overlap.pos))
unique(genotypic_tdt_results$overlap.pos)

ggplot(genotypic_tdt_results)+
  geom_point(aes(x=position,y=neglogp,color=overlap.pos$snp)) +
  labs(title="Genotypic TDT results, positions in low-p-value 10 overlap data in green",
       x="SNP position", y = "-logp")
```